src/controller/python/chip/yaml/parser.py

```diff
@@ -458,10 +428,37 @@ def _attribute_write_action_factory(self, item: dict, cluster: str):
         except ParsingError:
             return None
 
-    def execute_tests(self, dev_ctrl: ChipDeviceCtrl):
-        '''Executes parsed YAML tests.'''
+    def parse_and_execute_test(self, dev_ctrl: ChipDeviceCtrl):
```

Comment:
Please keep naming this as `execute_tests`. It should not matter whether there is some parsing/reparsing during execution. It's not a fundamental constraint, nor should it be. 

In essence, there are 2 passes: initial building of the tree of test items, and then, at execution, necessary substitutions. The fact that this can be implemented easier in the short term by iterating over the source in one go during execution is an implementation detail.

It should definitely be possible for the loading of the YAML file to yield most of the possible errors about structure, and only the "runtime" binding of variable (saveAs) to values could lead to errors, during the "evaluation"/"execution" phase.